diff -urN exim-4.21.org/src/daemon.c exim-4.21/src/daemon.c
--- exim-4.21.org/src/daemon.c	2003-08-18 00:04:37.000000000 +0200
+++ exim-4.21/src/daemon.c	2003-08-18 00:04:54.000000000 +0200
@@ -656,7 +656,7 @@
 */
 
 static BOOL
-check_special_case(int errno, ip_address_item *addresses, ip_address_item *ipa,
+check_special_case(int eno, ip_address_item *addresses, ip_address_item *ipa,
   BOOL back)
 {
 ip_address_item *ipa2;
@@ -667,7 +667,7 @@
 
 if (back)
   {
-  if (errno != EADDRINUSE || ipa->address[0] != 0) return FALSE;
+  if (eno != EADDRINUSE || ipa->address[0] != 0) return FALSE;
   for (ipa2 = addresses; ipa2 != ipa; ipa2 = ipa2->next)
     if (ipa2->address[1] == 0 && ipa2->port == ipa->port)
       return TRUE;
