diff -ruN exim-3.20.orig/OS/Makefile-Default exim-3.20/OS/Makefile-Default
--- exim-3.20.orig/OS/Makefile-Default	Sat Nov 25 17:02:45 2000
+++ exim-3.20/OS/Makefile-Default	Tue Dec  5 10:30:42 2000
@@ -25,14 +25,14 @@
 # found, it uses /usr/ucb/basename. This copes with Solaris 2 and Linux, both
 # of which come in different versions.
 
-BASENAME_COMMAND=/usr/bin/basename
+BASENAME_COMMAND=/bin/basename
 
 
 # Some of the following commands live in different places in different OS. We
 # include them all here for generality.
 
-CHOWN_COMMAND=/usr/bin/chown
-CHGRP_COMMAND=/usr/bin/chgrp
+CHOWN_COMMAND=/bin/chown
+CHGRP_COMMAND=/bin/chgrp
 MV_COMMAND=/bin/mv
 RM_COMMAND=/bin/rm
 
@@ -142,7 +142,7 @@
 # parameters for include and library directories that may be needed for IPv6
 # on some systems.
 
-# HAVE_IPV6=YES
+HAVE_IPV6=YES
 # IPV6_INCLUDE=-I /usr/ipv6/include
 # IPV6_LIBS=-L/usr/ipv6/libs -linet6
 
