--- exim-4.93/src/EDITME.org	2019-12-11 07:24:16.057781209 +0100
+++ exim-4.93/src/EDITME	2019-12-11 07:24:21.077926862 +0100
@@ -100,7 +100,7 @@
 # /usr/local/sbin. The installation script will try to create this directory,
 # and any superior directories, if they do not exist.
 
-BIN_DIRECTORY=/usr/exim/bin
+BIN_DIRECTORY=/usr/bin
 
 
 #------------------------------------------------------------------------------
@@ -116,7 +116,7 @@ BIN_DIRECTORY=/usr/exim/bin
 # don't exist. It will also install a default runtime configuration if this
 # file does not exist.
 
-CONFIGURE_FILE=/usr/exim/configure
+CONFIGURE_FILE=/etc/mail/exim.conf
 
 # It is possible to specify a colon-separated list of files for CONFIGURE_FILE.
 # In this case, Exim will use the first of them that exists when it is run.
@@ -133,7 +133,8 @@ CONFIGURE_FILE=/usr/exim/configure
 # deliveries. (Local deliveries run as various non-root users, typically as the
 # owner of a local mailbox.) Specifying these values as root is not supported.
 
-EXIM_USER=
+EXIM_UID=79
+EXIM_GID=79
 
 # If you specify EXIM_USER as a name, this is looked up at build time, and the
 # uid number is built into the binary. However, you can specify that this
@@ -211,11 +211,11 @@ SPOOL_DIRECTORY=/var/spool/exim
 # If you are building with TLS, the library configuration must be done:
 
 # Uncomment this if you are using OpenSSL
-# USE_OPENSSL=yes
+USE_OPENSSL=yes
 # Uncomment one of these settings if you are using OpenSSL; pkg-config vs not
 # and an optional location.
-# USE_OPENSSL_PC=openssl
-# TLS_LIBS=-lssl -lcrypto
+USE_OPENSSL_PC=openssl
+TLS_LIBS=-lssl -lcrypto
 # TLS_LIBS=-L/usr/local/openssl/lib -lssl -lcrypto
 
 # Uncomment this if you are using GnuTLS
@@ -317,7 +318,7 @@ ROUTER_REDIRECT=yes
 
 # This one is very special-purpose, so is not included by default.
 
-# ROUTER_IPLOOKUP=yes
+ROUTER_IPLOOKUP=yes
 
 
 #------------------------------------------------------------------------------
@@ -338,7 +339,7 @@ TRANSPORT_SMTP=yes
 # This one is special-purpose, and commonly not required, so it is not
 # included by default.
 
-# TRANSPORT_LMTP=yes
+TRANSPORT_LMTP=yes
 
 
 #------------------------------------------------------------------------------
@@ -347,9 +348,9 @@ TRANSPORT_SMTP=yes
 # MBX, is included only when requested. If you do not know what this is about,
 # leave these settings commented out.
 
-# SUPPORT_MAILDIR=yes
-# SUPPORT_MAILSTORE=yes
-# SUPPORT_MBX=yes
+SUPPORT_MAILDIR=yes
+SUPPORT_MAILSTORE=yes
+SUPPORT_MBX=yes
 
 
 #------------------------------------------------------------------------------
@@ -407,8 +408,8 @@ LOOKUP_DBM=yes
 LOOKUP_LSEARCH=yes
 LOOKUP_DNSDB=yes
 
-# LOOKUP_CDB=yes
-# LOOKUP_DSEARCH=yes
+LOOKUP_CDB=yes
+LOOKUP_DSEARCH=yes
 # LOOKUP_JSON=yes
 # LOOKUP_LDAP=yes
 # LOOKUP_LMDB=yes
@@ -417,7 +418,7 @@ LOOKUP_DNSDB=yes
 # LOOKUP_NIS=yes
 # LOOKUP_NISPLUS=yes
 # LOOKUP_ORACLE=yes
-# LOOKUP_PASSWD=yes
+LOOKUP_PASSWD=yes
 # LOOKUP_PGSQL=yes
 # LOOKUP_PSL=yes
 # LOOKUP_REDIS=yes
@@ -500,7 +500,7 @@ SUPPORT_DANE=yes
 # files are defaulted in the OS/Makefile-Default file, but can be overridden in
 # local OS-specific make files.
 
-# EXIM_MONITOR=eximon.bin
+EXIM_MONITOR=eximon.bin
 
 
 #------------------------------------------------------------------------------
@@ -509,7 +510,7 @@ SUPPORT_DANE=yes
 # and the MIME ACL. Please read the documentation to learn more about these
 # features.
 
-# WITH_CONTENT_SCAN=yes
+WITH_CONTENT_SCAN=yes
 
 # If you have content scanning you may wish to only include some of the scanner
 # interfaces.  Uncomment any of these lines to remove that code.
@@ -657,7 +657,7 @@ DISABLE_MAL_MKS=yes
 
 # Uncomment the following lines to add SRS (Sender Rewriting Scheme) support.
 # This needs no library.
-# SUPPORT_SRS=yes
+SUPPORT_SRS=yes
 
 # Uncomment the following to remove support for the ESMTP extension "WELLKNOWN"
 # DISABLE_WELLKNOWN=yes
@@ -751,7 +751,7 @@ FIXED_NEVER_USERS=root
 # the TRUSTED_CONFIG_LIST file, then root privileges are not dropped by Exim.
 
 # TRUSTED_CONFIG_LIST=/usr/exim/trusted_configs
-
+TRUSTED_CONFIG_LIST=/etc/exim/trusted_configs
 
 #------------------------------------------------------------------------------
 # Uncommenting this option disables the use of the -D command line option,
@@ -880,18 +880,18 @@ TRUSTED_CONFIG_LIST=/etc/exim/trusted_co
 # core exim build.  This gets them linked with the module instead.
 # The heimdal does build but we have no test coverage so it is not know to work.
 
-# AUTH_CRAM_MD5=yes
+AUTH_CRAM_MD5=yes
 # AUTH_CYRUS_SASL=yes
-# AUTH_DOVECOT=yes
+AUTH_DOVECOT=yes
 # AUTH_EXTERNAL=yes
 # AUTH_GSASL=yes
 # AUTH_GSASL_PC=libgsasl
 # AUTH_HEIMDAL_GSSAPI=yes
 # AUTH_HEIMDAL_GSSAPI_PC=heimdal-gssapi
 # AUTH_HEIMDAL_GSSAPI_PC=heimdal-gssapi heimdal-krb5
-# AUTH_PLAINTEXT=yes
-# AUTH_SPA=yes
-# AUTH_TLS=yes
+AUTH_PLAINTEXT=yes
+AUTH_SPA=yes
+AUTH_TLS=yes
 
 # Heimdal through 1.5 required pkg-config 'heimdal-gssapi'; Heimdal 7.1
 # requires multiple pkg-config files to work with Exim, so the second example
@@ -787,7 +788,7 @@ FIXED_NEVER_USERS=root
 # one that is set in the headers_charset option. The default setting is
 # defined by this setting:
 
-HEADERS_CHARSET="ISO-8859-1"
+HEADERS_CHARSET="UTF-8"
 
 # If you are going to make use of $header_xxx expansions in your configuration
 # file, or if your users are going to use them in filter files, and the normal
@@ -883,7 +898,7 @@ HEADERS_CHARSET="ISO-8859-1"
 # Once you have done this, "make install" will build the info files and
 # install them in the directory you have defined.
 
-# INFO_DIRECTORY=/usr/share/info
+INFO_DIRECTORY=/usr/share/info
 
 
 #------------------------------------------------------------------------------
@@ -896,7 +911,7 @@ HEADERS_CHARSET="ISO-8859-1"
 # %s. This will be replaced by one of the strings "main", "panic", or "reject"
 # to form the final file names. Some installations may want something like this:
 
-# LOG_FILE_PATH=/var/log/exim_%slog
+LOG_FILE_PATH=/var/log/exim/%s.log
 
 # which results in files with names /var/log/exim_mainlog, etc. The directory
 # in which the log files are placed must exist; Exim does not try to create
@@ -945,7 +960,7 @@ EXICYCLOG_MAX=10
 # files. Both the name of the command and the suffix that it adds to files
 # need to be defined here. See also the EXICYCLOG_MAX configuration.
 
-COMPRESS_COMMAND=/usr/bin/gzip
+COMPRESS_COMMAND=/bin/gzip
 COMPRESS_SUFFIX=gz
 
 
@@ -960,7 +975,7 @@ COMPRESS_SUFFIX=gz
 # ZCAT_COMMAND=zcat
 #
 # Or specify the full pathname:
-ZCAT_COMMAND=/usr/bin/zcat
+ZCAT_COMMAND=/bin/zcat
 
 #------------------------------------------------------------------------------
 # Compiling in support for embedded Perl: If you want to be able to
@@ -1100,7 +1100,7 @@ ZCAT_COMMAND=/bin/zcat
 # Perl costs quite a lot of resources. Only do this if you really need it.
 #
 
-# EXIM_PERL=perl.o
+EXIM_PERL=perl.o
 
 # For a dynamic module build add also SUPPORT_PERL=2 and SUPPORT_PAM_(INCLUED,LIBS)
 #SUPPORT_PERL=2
@@ -978,7 +993,7 @@ ZCAT_COMMAND=/usr/bin/zcat
 # that the local_scan API is made available by the linker. You may also need
 # to add -ldl to EXTRALIBS so that dlopen() is available to Exim.
 
-# EXPAND_DLFUNC=yes
+EXPAND_DLFUNC=yes
 
 
 #------------------------------------------------------------------------------
@@ -1127,11 +1127,12 @@ EXPAND_DLFUNC=yes
 #
 # For a dynamic module build add SUPPORT_PAM=2 and SUPPORT_PAM_LIBS=-lpam
 
-# SUPPORT_PAM=yes
+SUPPORT_PAM=yes
 
 # You probably need to add -lpam to EXTRALIBS, and in some releases of
 # GNU/Linux -ldl is also needed.
 
+EXTRALIBS=-lpam -ldl
 
 #------------------------------------------------------------------------------
 # Proxying.
@@ -1096,7 +1112,7 @@ ZCAT_COMMAND=/usr/bin/zcat
 # group. Once you have installed saslauthd, you should arrange for it to be
 # started by root at boot time.
 
-# CYRUS_SASLAUTHD_SOCKET=/var/state/saslauthd/mux
+CYRUS_SASLAUTHD_SOCKET=/var/lib/sasl2/mux
 
 
 #------------------------------------------------------------------------------
@@ -1163,7 +1179,7 @@ SYSTEM_ALIASES_FILE=/etc/aliases
 # is "yes", as well as supporting line editing, a history of input lines in the
 # current run is maintained.
 
-# USE_READLINE=yes
+USE_READLINE=yes
 
 # You may need to add -ldl to EXTRALIBS when you set USE_READLINE=yes.
 # Note that this option adds to the size of the Exim binary, because the
@@ -1253,7 +1269,7 @@ EXIM_TMPDIR="/tmp"
 # period (for example, /usr/exim/configure.host.in.some.domain). If this file
 # does not exist, then the bare configuration file name is tried.
 
-# CONFIGURE_FILE_USE_NODE=yes
+CONFIGURE_FILE_USE_NODE=yes
 
 
 #------------------------------------------------------------------------------
@@ -1432,7 +1448,7 @@ EXIM_TMPDIR="/tmp"
 # be changed here. The default is 0640 so that information from the spool is
 # available to anyone who is a member of the Exim group.
 
-# SPOOL_MODE=0640
+SPOOL_MODE=0640
 
 
 #------------------------------------------------------------------------------
@@ -1478,7 +1494,7 @@ EXIM_TMPDIR="/tmp"
 # feature should only be used in very exceptional circumstances. YOU HAVE BEEN
 # WARNED.
 
-# ENABLE_DISABLE_FSYNC=yes
+ENABLE_DISABLE_FSYNC=yes
 
 #------------------------------------------------------------------------------
 # For development, add this to include code to time various stages and report.
