diff -urN exim-4.42.org/src/exim.c exim-4.42/src/exim.c
--- exim-4.42.org/src/exim.c	2004-09-21 15:04:23.506142125 +0200
+++ exim-4.42/src/exim.c	2004-09-21 15:08:08.497942908 +0200
@@ -3551,7 +3551,7 @@
 delays are in evidence. Save the home directory for use in filter testing
 (only). */
 
-for (i = 1; i <= 10; i++)
+for (i = 1; i <= (finduser_retries + 1); i++)
   {
   if ((pw = getpwuid(real_uid)) != NULL)
     {
@@ -3620,7 +3620,8 @@
 
     break;
     }
-  sleep(1);
+  if (finduser_retries)
+      sleep(1);
   }
 
 /* If we cannot get a user login, log the incident and give up, unless the
