diff -urN eximorg/exim4.conf exim/exim4.conf
--- eximorg/exim4.conf	2015-01-13 09:39:15.705251985 +0100
+++ exim/exim4.conf	2015-01-13 09:43:24.017910803 +0100
@@ -145,6 +145,8 @@ acl_smtp_data = acl_check_data
 # Disable TLS
 tls_advertise_hosts =
 
+openssl_options = +no_sslv2 +no_sslv3
+
 # Specify the location of the Exim server's TLS certificate and private key.
 # The private key must not be encrypted (password protected). You can put
 # the certificate and private key in the same file, in which case you only

