diff --git a/src/transports/smtp.c b/src/transports/smtp.c
index 59abb9ef8..ad648755f 100644
--- a/src/transports/smtp.c
+++ b/src/transports/smtp.c
@@ -5059,7 +5053,8 @@ if (max_received_linelength > ob->message_linelength_limit)
       addr->transport_return = PENDING_DEFER;
 
   set_errno_nohost(addrlist, ERRNO_SMTPFORMAT,
-    US"message has lines too long for transport", FAIL, TRUE, &now);
+    string_sprintf("message has lines too long for transport (received line length %d while limit is %d)",
+      max_received_linelength, ob->message_linelength_limit), FAIL, TRUE, &now);
   goto END_TRANSPORT;
   }
 

