From 740f36d42bb3f61fdbaf53a68380a9a96096c229 Mon Sep 17 00:00:00 2001
From: Jeremy Harris <jgh146exb@wizmail.org>
Date: Wed, 25 Mar 2015 17:11:56 +0000
Subject: [PATCH] Fix recent-openssl compile

---
 src/tls-openssl.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/src/tls-openssl.c b/src/tls-openssl.c
index 96ac72c..65d6089 100644
--- a/src/tls-openssl.c
+++ b/src/tls-openssl.c
@@ -380,7 +380,7 @@ else
      	/* client, wanting hostname check */
     {
 
-# if EXIM_HAVE_OPENSSL_CHECKHOST
+# ifdef EXIM_HAVE_OPENSSL_CHECKHOST
 #  ifndef X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS
 #   define X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS 0
 #  endif
@@ -394,7 +394,8 @@ else
     while ((name = string_nextinlist(&list, &sep, NULL, 0)))
       if ((rc = X509_check_host(cert, name, 0,
 		  X509_CHECK_FLAG_NO_PARTIAL_WILDCARDS
-		  | X509_CHECK_FLAG_SINGLE_LABEL_SUBDOMAINS)))
+		  | X509_CHECK_FLAG_SINGLE_LABEL_SUBDOMAINS,
+		  NULL)))
 	{
 	if (rc < 0)
 	  {
-- 
1.9.1

