--- exim-4.95.org/src/tls-openssl.c	2021-09-28 08:24:46.000000000 +0000
+++ exim-4.95/src/tls-openssl.c	2021-10-06 06:27:39.949989669 +0000
@@ -232,12 +232,7 @@
   { US"no_tlsv1", SSL_OP_NO_TLSv1 },
 #endif
 #ifdef SSL_OP_NO_TLSv1_1
-# if SSL_OP_NO_TLSv1_1 == 0x00000400L
-  /* Error in chosen value in 1.0.1a; see first item in CHANGES for 1.0.1b */
-#  warning OpenSSL 1.0.1a uses a bad value for SSL_OP_NO_TLSv1_1, ignoring
-# else
   { US"no_tlsv1_1", SSL_OP_NO_TLSv1_1 },
-# endif
 #endif
 #ifdef SSL_OP_NO_TLSv1_2
   { US"no_tlsv1_2", SSL_OP_NO_TLSv1_2 },
