commit 4f252517dc0e8eda6c1eb1031ac0bcd29b6a7a93
Author: Arkadiusz Miśkiewicz <arekm@maven.pl>
Date:   Thu Jan 23 08:51:01 2020 +0100

    Add support for zstd compressed .zst files (Bug 2515)

diff --git a/src/src/exigrep.src b/src/src/exigrep.src
index afd3538b6..835766fb4 100644
--- a/src/src/exigrep.src
+++ b/src/src/exigrep.src
@@ -159,7 +159,8 @@ my $compressors = {
   gz   => { cmd => 'zcat',  args => '' },
   bz2  => { cmd => 'bzcat', args => '' },
   xz   => { cmd => 'xzcat', args => '' },
-  lzma => { cmd => 'lzma',  args => '-dc' }
+  lzma => { cmd => 'lzma',  args => '-dc' },
+  zst  => { cmd => 'zstdcat', args => '' },
 };
 my $csearch = 0;
 
@@ -313,8 +314,8 @@ given host, for example.
 
 If no file names are given on the command line, the standard input is read.
 
-For known file extensions indicating compression (F<.gz>, F<.bz2>, F<.xz>, and F<.lzma>)
-a suitable de-compressor is used, if available.
+For known file extensions indicating compression (F<.gz>, F<.bz2>, F<.xz>,
+F<.lzma>, and F<.zst>) a suitable de-compressor is used, if available.
 
 The output is sent through a pager if a terminal is connected to STDOUT. As
 pager are considered: C<$ENV{PAGER}>, C<less>, C<more>.
